controller/cache/node.go

```diff
@@ -0,0 +1,76 @@
+package cache
+
+import (
+	appv1 "github.com/argoproj/argo-cd/pkg/apis/application/v1alpha1"
+	"github.com/argoproj/argo-cd/util/kube"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type node struct {
+	resourceVersion string
+	ref             v1.ObjectReference
+	ownerRefs       []metav1.OwnerReference
+	children        map[string]*node
+	parents         map[string]*node
+	tags            []string
+	appName         string
+	resource        *unstructured.Unstructured
+}
+
+// returns true if node has application label and has no parents
+func (o *node) isAppRoot() bool {
+	return o.appName != "" && len(o.parents) == 0
+}
+
+func (o *node) resourceKey() string {
+	return kube.FormatResourceKey(o.ref.GroupVersionKind().Group, o.ref.Kind, o.ref.Namespace, o.ref.Name)
+}
```

Comment:
I think it would be better if we could use structs as resource keys. Being able to iterate the node map and filter by namespace, would be trivial using a struct, but harder if we need to split based on a string convention.

Reply:
Nice suggestion. Done